File: 871-keys-and-rooms/keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        adj = defaultdict(list)
        for i, keys in enumerate(rooms):
            adj[i].extend(keys)
            
        stk = [0]
        visited = set()
        while(stk):
            cur = stk.pop()
            visited.add(cur)
            for nei in adj[cur]:
                if nei not in visited:
                    stk.append(nei)
        
        return len(visited) == len(rooms)
        

File: 871-keys-and-rooms/keys-and-rooms.cpp
class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        stack<int> stk;
        set<int> visited;
        stk.push(0);
        while(!stk.empty()) {
            int cur = stk.top();
            stk.pop();
            visited.insert(cur);
            for(auto key: rooms[cur]) {
                if(visited.find(key) == visited.end()) {
                    stk.push(key);
                }
            }
        }
        return visited.size() == rooms.size();
    }
};